### ISSUE TITLE
memory leak when accessing matrix with [][]

### ISSUE THREAD

**Issue by knuedd** [dash-project/dash#24]
In an example application I noticed a huge memory consumption. The following is a minimal example that demonstrates this behavior. Like this, it exceeds my 8 gig memory and is killed. **Make sure to disable your swap before testing.** 

```
#include <unistd.h>
#include <iostream>
#include <libdash.h>

int main( int argc, char* argv[]) {

    dash::init(&argc, &argv);

    size_t myid= dash::myid();
    size_t team_size= dash::Team::All().size();
    dash::TeamSpec<2> teamspec_2d(team_size, 1);
    teamspec_2d.balance_extents();

    uint32_t w= 8000;
    uint32_t h= 8000;

    dash::Matrix<uint32_t, 2> matrix( dash::SizeSpec<2>( w, h ),
        dash::DistributionSpec<2>( dash::BLOCKED, dash::BLOCKED ),
        dash::Team::All(), teamspec_2d );

    if ( 0 == myid ) {

        auto it= matrix.begin(); // only needed for NOLEAK case
        for(size_t i = 0; i < w; ++i) {
            for(size_t k = 0; k < h; ++k) {

#define LEAK
#ifdef LEAK
                matrix[i][k] = 100; // me makes mighty memory leak
# else /* LEAK*/
                *it= 100; ++it;
#endif /* LEAK */
            }
        }
    }

    return 0;
}
```

Change '#define LEAK' into '#define NOLEAK' to get a perfectly working code. This makes me think that there is no other cause of a memory leak. Did I confuse something? Anything that should not be used like this in a regular code?

**Comment by fuchsto** [comment]
Yes, just confirmed this has been fixed by @devreal in #22

**Comment by knuedd** [comment]
Thanks for the quick reaction. Unfortunately, my little test code above is now killed during dash::init(). This is what happens:

```
(gdb) run
Starting program: /home/knue/Private/prog/dash_tutorial/leak 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff4c62700 (LWP 13340)]
[New Thread 0x7fffeffff700 (LWP 13341)]
[    0 ERROR ] [ 13331 ] dart_locality.c          :62   !!! DART: dart_domain_locality: dart__base__locality__domain failed for domain tag '' -> (3)
[    0 ERROR ] [ 13331 ] Locality.cc              :110  | dash::exception::RuntimeError                | [ Unit 0 ] Locality::init(): dart_domain_team_locality failed for domain '' 
terminate called after throwing an instance of 'dash::exception::RuntimeError'
  what():  [ Unit 0 ] Locality::init(): dart_domain_team_locality failed for domain ''

Thread 1 "leak" received signal SIGABRT, Aborted.
0x00007ffff6ca4418 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
54      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) where
#0  0x00007ffff6ca4418 in __GI_raise (sig=sig@entry=6)
    at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007ffff6ca601a in __GI_abort () at abort.c:89
#2  0x00007ffff780384d in __gnu_cxx::__verbose_terminate_handler() ()
   from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007ffff78016b6 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff7801701 in std::terminate() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff7801919 in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x000000000041599d in dash::util::Locality::init() ()
#7  0x0000000000404f34 in main (argc=1, argv=0x7fffffffdc28) at leak.cpp:7
(gdb) 
```

**Comment by knuedd** [comment]
Hm, it seems the memory leak was fixed already in 9cfe34d955374cddec2e897d2789b791df4d276b.

**Comment by fuchsto** [comment]
Yes, in pull request #22 
I think the initialization has a bug if hwloc and / or PAPI are not enabled.
Could you attach your build log?

**Comment by knuedd** [comment]
Indeed, I didn't have PAPI and hwloc. Here is the cmake log which should be enough. Or do you need the entire build output?

```
 -- Loading environment configuration from /home/knue/prog/dash_github/config/default.cmake
-- Searching for PAPI in path PAPI_PREFIX-NOTFOUND
-- Could NOT find PAPI (missing:  PAPI_LIBRARIES PAPI_INCLUDE_DIRS) 
-- Searching for hwloc library in path HWLOC_PREFIX-NOTFOUND
-- Could NOT find HWLOC (missing:  HWLOC_LIBRARIES HWLOC_INCLUDE_DIRS) 
-- likwid includes:  /usr/local/include
-- likwid libraries: /usr/local/lib/liblikwid.so
-- Searching for NUMA library
-- Could NOT find NUMA (missing:  NUMA_LIBRARIES) 
-- Searching MKL Intel64
-- Intel(R) MKL could not be found.
-- (ENV) MKLROOT:       
-- (ENV) INTELROOT:     
-- (ENV) SCALAPACK_LIB: 
-- Intel(R) MKL could not be found.
-- (ENV) MKLROOT:       
-- (ENV) INTELROOT:     
-- (ENV) SCALAPACK_LIB: 
-- Searching for IPM library in path IPM_PREFIX-NOTFOUND
-- Could NOT find IPM (missing:  IPM_LIBRARIES IPM_INCLUDE_DIRS) 
-- Searching for Plasma library in path PLASMA_PREFIX-NOTFOUND
-- Could NOT find PLASMA (missing:  PLASMA_LIBRARIES PLASMA_INCLUDE_DIRS) 
-- Trying to find HDF5 using cmake package
-- Could NOT find HDF5 (missing:  HDF5_LIBRARIES HDF5_INCLUDE_DIRS) 
-- HDF5 cmake module provides only serial version
-- HDF5 package not found, try to find libs
-- Could NOT find HDF5 (missing:  HDF5_LIBRARIES HDF5_INCLUDE_DIRS) 
-- A library with BLAS API not found. Please specify library location.
-- A library with BLAS API not found. Please specify library location.
-- LAPACK requires BLAS
-- A library with LAPACK API not found. Please specify library location.
-- LAPACK not found
-- Could NOT find SCALAPACK (missing:  SCALAPACK_LIBRARIES SCALAPACK_INCLUDE_DIRS) 
-- CC  flags (Debug):   -g  -mtune=native -march=native -mfpmath=sse --std=c99 -fopenmp -Wall -Wextra -pedantic -Wno-unused-function -Wno-missing-braces -Wno-format -Wno-unused-parameter -O0 -DDASH_DEBUG -ggdb3
-- CXX flags (Debug):   -g  -mtune=native -march=native -mfpmath=sse --std=c++11 -fopenmp -Wall -Wextra -pedantic -Wno-unused-function -Wno-missing-braces -Wno-format -Wno-unused-parameter -O0 -DDASH_DEBUG -ggdb3 -rdynamic
-- CC  flags (Release): -O3 -DNDEBUG  -mtune=native -march=native -mfpmath=sse --std=c99 -fopenmp -Wall -Wextra -pedantic -Wno-unused-function -Wno-missing-braces -Wno-format -Wno-unused-parameter -Ofast -DDASH_RELEASE
-- CXX flags (Release): -O3 -DNDEBUG  -mtune=native -march=native -mfpmath=sse --std=c++11 -fopenmp -Wall -Wextra -pedantic -Wno-unused-function -Wno-missing-braces -Wno-format -Wno-unused-parameter -Ofast -DDASH_RELEASE
-- MPI include path:         /opt/open-mpi/include
-- MPI libraries:
--     /opt/open-mpi/lib/libmpi.so
-- MPI link flags:
--     -Wl,-rpath  -Wl,/opt/open-mpi/lib  -Wl,--enable-new-dtags
-- DART additional compile flags:
--     -DDART
--     -DDART_MPI_IMPL_ID=""
--     -DDART_MPI_ENABLE_UNIFIED_MEMORY_MODEL
-- DART additional libraries:
--     rt
-- DASH additional compile flags:
--     -fopenmp
--     -DDASH
--     -DDASH_ENABLE_DEFAULT_INDEX_TYPE_LONG
--     -DDASH_ENABLE_UNIFIED_MEMORY_MODEL
--     -DDASH_ENABLE_TEST_LOGGING
--     -DDASH_ENV_HOST_SYSTEM_ID='default'
-- DASH additional libraries:
--     rt
--     pthread
-- Building DASH library    dash-mpi
-- Downloading GTest from official repository
-- DASH version 0.3.0
-- DART interface version 3.2
-- -----------------------------------------------------------
-- Install prefix:           (INSTALL_PREFIX)                 /opt/dash/
-- Build type:               (CMAKE_BUILD_TYPE)               Release
-- Host system identifier:   (ENVIRONMENT_TYPE)               default
-- Strict compiler warnings: (ENABLE_COMPILER_WARNINGS)       ON
-- Debug log messages:       (ENABLE_LOGGING)                 OFF
-- Trace log messages:       (ENABLE_TRACE_LOGGING)           OFF
-- DART log messages:        (ENABLE_DART_LOGGING)            OFF
-- Runtime assertions:       (ENABLE_ASSERTIONS)              OFF
-- Unified RMA memory model: (ENABLE_UNIFIED_MEMORY_MODEL)    ON
-- MPI shared windows:       (ENABLE_SHARED_WINDOWS)          ON
-- Default index type long:  (ENABLE_DEFAULT_INDEX_TYPE_LONG) ON
-- libnuma support:          (ENABLE_LIBNUMA)                 ON
-- hwloc support:            (ENABLE_HWLOC)                   ON
-- PAPI support:             (ENABLE_PAPI)                    ON
-- MKL support:              (ENABLE_MKL)                     ON
-- BLAS support:             (ENABLE_BLAS)                    ON
-- LAPACK support:           (ENABLE_LAPACK)                  ON
-- ScaLAPACK support:        (ENABLE_SCALAPACK)               ON
-- PLASMA support:           (ENABLE_PLASMA)                  ON
-- HDF5 support:             (ENABLE_HDF5)                    ON
-- Enabled DART backends:    (DART_IMPLEMENTATIONS)           mpi
-- C   compiler id:          GNU
-- C++ compiler id:          GNU
-- MPI implementation:       
-- PAPI                      not found
-- MKL                       not found
-- BLAS                      not found
-- hwloc                     not found
-- likwid                    disabled
-- LAPACK                    not found
-- ScaLAPACK                 not found
-- PLASMA                    not found
-- HDF5                      not found
-- libnuma                   not found
-- Configuring done
-- Generating done
-- Build files have been written to: /home/knue/prog/dash_github/build
-- Cache values
BLACS_LIBRARY:FILEPATH=BLACS_LIBRARY-NOTFOUND
BUILD_COVERAGE_TESTS:BOOL=OFF
BUILD_DOCS:BOOL=OFF
BUILD_EXAMPLES:BOOL=ON
BUILD_TESTS:BOOL=ON
CC_GDB_FLAG:STRING=-ggdb3
CC_STD_FLAG:STRING=--std=c99
CMAKE_BUILD_TYPE:STRING=
CMAKE_INSTALL_PREFIX:PATH=/usr/local
CXX_GDB_FLAG:STRING=-ggdb3 -rdynamic
CXX_STD_FLAG:STRING=--std=c++11
DART_IF_VERSION:STRING=3.2
DART_IMPLEMENTATIONS:STRING=mpi
DART_IMPLEMENTATION_MPI_ENABLED:BOOL=ON
DASH_DART_BASE_INCLUDE_DIR:PATH=/home/knue/prog/dash_github/dart-impl/base/include
DASH_DART_IF_INCLUDE_DIR:PATH=/home/knue/prog/dash_github/dart-if/v3.2/include
DASH_ENV_HOST_SYSTEM_ID:STRING=default
DASH_PLATFORM_IS_POSIX:BOOL=TRUE
DASH_PLATFORM_NAME:STRING=Linux
DASH_VERSION:STRING=0.3.0
DASH_VERSIONED_PROJECT_NAME:STRING=dash-0.3.0
ENABLE_ASSERTIONS:BOOL=OFF
ENABLE_BLAS:BOOL=ON
ENABLE_COMPILER_WARNINGS:BOOL=ON
ENABLE_DART_LOGGING:BOOL=OFF
ENABLE_DEFAULT_INDEX_TYPE_LONG:BOOL=ON
ENABLE_HDF5:BOOL=ON
ENABLE_HWLOC:BOOL=ON
ENABLE_HWLOC_PCI:BOOL=ON
ENABLE_LAPACK:BOOL=ON
ENABLE_LIBNUMA:BOOL=ON
ENABLE_LIKWID:BOOL=OFF
ENABLE_LOGGING:BOOL=OFF
ENABLE_MKL:BOOL=ON
ENABLE_PAPI:BOOL=ON
ENABLE_PLASMA:BOOL=ON
ENABLE_SCALAPACK:BOOL=ON
ENABLE_SHARED_WINDOWS:BOOL=ON
ENABLE_TEST_LOGGING:BOOL=ON
ENABLE_TRACE_LOGGING:BOOL=OFF
ENABLE_UNIFIED_MEMORY_MODEL:BOOL=ON
ENVIRONMENT_CONFIG_PATH:STRING=
ENVIRONMENT_TYPE:STRING=default
GTEST_INCLUDE_PATH:STRING=
GTEST_LIBRARY_PATH:STRING=
HDF5_DIR:PATH=HDF5_DIR-NOTFOUND
HDF5_FOUND:BOOL=OFF
HWLOC_PREFIX:PATH=HWLOC_PREFIX-NOTFOUND
IOMP5_LIBRARY:FILEPATH=IOMP5_LIBRARY-NOTFOUND
IPM_PREFIX:PATH=IPM_PREFIX-NOTFOUND
LIKWID_PREFIX:PATH=/usr/local
MATH_LIBRARY:FILEPATH=MATH_LIBRARY-NOTFOUND
MPI_EXTRA_LIBRARY:STRING=MPI_EXTRA_LIBRARY-NOTFOUND
MPI_FOUND:BOOL=FALSE
MPI_LIBRARY:FILEPATH=/opt/open-mpi/lib/libmpi.so
PAPI_PREFIX:PATH=PAPI_PREFIX-NOTFOUND
PLASMA_PREFIX:PATH=PLASMA_PREFIX-NOTFOUND
PTHREAD_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libpthread.so
SCALAPACK_LIBRARY:FILEPATH=SCALAPACK_LIBRARY-NOTFOUND
WARNINGS_AS_ERRORS:BOOL=OFF
```

**Comment by fuchsto** [comment]
Thank you @knuedd, the CMake config preamble is exactly what I needed. Could reproduce the defect easily. Closing this issue as the original defect is fixed, created separate issue #34 for the defect in initialization of DART.
